### PR TITLE
Add dialect support 

### PR DESCRIPTION
**Pull request by DvirDukhan**
This PR introduces new dialect support to FT.SEARCH and FT.AGGREGATE commands, as well as FT.CONFIG and module load time flags.
This PR introduces two dialect versions `1` and `2`. The first one is the dialect being used in RediSearch 2.2 and the latter is RediSearch 2.4 dialect that includes parameters, VSS, and logical bug fixes.

By default, RediSearch 2.4 will be loaded with dialect version 1

1. dialect control via configuration load time flag
`redis-server --loadmodule path/to/redisearch.so DEFAULT_DIALECT 2`
This will set the default dialect for every command to be 2, unless overridden by a command argument
2. dialect control via FT.CONFIG
`FT.CONFIG SET DEFAULT_DIALECT 2`
This will allow setting the default dialect during runtime
3. dialect control on command level
`FT.SEARCH idx "hello" DIALECT 2`
This will execute query processing with dialect version 2, regardless the default dialect version set during module loading/FT.CONFIG call

### CODE REVIEW

File: coord/src/module.c [hunk 1]
```diff
@@ -437,22 +439,21 @@ static int rscParseProfile(searchRequestCtx *req, RedisModuleString **argv) {
...
-                         .status = &status,
+                         .status = status,
 #ifdef PARSER_DEBUG
                          .trace_log = NULL
 #endif
   };
   QueryNode* queryNode = RSQuery_ParseRaw(&qpCtx);
-  if(status.code != 0 ) {
+  if(status->code != 0 ) {
     //fail.
```
Comment (MeirShpilraien) [inline coord/src/module.c:457]: return?

File: coord/src/module.c [hunk 2]
```diff
@@ -595,10 +596,45 @@ searchRequestCtx *rscParseRequest(RedisModuleString **argv, int argc) {
...
+        return NULL;	
+      }
+      if(dialect > MAX_DIALECT_VERSION) {
+        QueryError_SetErrorFmt(status, QUERY_EPARSEARGS, "DIALECT requires a non negative integer >=1 and <= %u", MAX_DIALECT_VERSION);	
+        free(req);
+        return NULL;
+    }
+
+    if(dialect >= 2) {
+        // Note: currently there is only one single case. For extending those cases we should use a trie here.
```
Comment (MeirShpilraien) [inline coord/src/module.c:622]: Is the coverage right? Is it not covered?

File: src/redisearch_api.c [hunk 3]
```diff
@@ -498,7 +498,7 @@ static RS_ApiIter* handleIterCommon(IndexSpec* sp, QueryInput* input, char** err
   RS_ApiIter* it = rm_calloc(1, sizeof(*it));
 
   if (input->qtype == QUERY_INPUT_STRING) {
-    if (QAST_Parse(&it->qast, &sctx, &options, input->u.s.qs, input->u.s.n, &status) !=
+    if (QAST_Parse(&it->qast, &sctx, &options, input->u.s.qs, input->u.s.n, RSGlobalConfig.defaultDialectVersion, &status) !=
```
Comment (MeirShpilraien) [inline src/redisearch_api.c:501]: Should be always 1. It is not possible to set it from LLAPI. We should expose a new API that gets the dialect version and deprecate this one.

File: src/module.c [hunk 4]
```diff
@@ -126,7 +126,7 @@ int SpellCheckCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
   const char **includeDict = NULL, **excludeDict = NULL;
   RSSearchOptions opts = {0};
   QueryAST qast = {0};
-  int rc = QAST_Parse(&qast, sctx, &opts, rawQuery, len, &status);
+  int rc = QAST_Parse(&qast, sctx, &opts, rawQuery, len, RSGlobalConfig.defaultDialectVersion, &status);
```
Comment (MeirShpilraien) [inline src/module.c:129]: We need to add it here also as a parameter

**Review by MeirShpilraien — COMMENTED** [review]
:+1: 
Some comments

**Review by oshadmi — COMMENTED** [review]
Some comments

File: coord/src/module.c [hunk 5]
```diff
@@ -595,9 +596,34 @@ searchRequestCtx *rscParseRequest(RedisModuleString **argv, int argc) {
...
+        return NULL;	
+      }
+      if(dialect > MAX_DIALECT_VERSION) {
+        QueryError_SetErrorFmt(status, QUERY_EPARSEARGS, "DIALECT requires a non negative integer >=1 and <= %u", MAX_DIALECT_VERSION);	
+        free(req);
+        return NULL;
+    }
+  }
+
+  if(dialect >= 2) {
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply coord/src/module.c:622]: future proof?

File: coord/src/module.c [hunk 6]
```diff
@@ -595,9 +596,34 @@ searchRequestCtx *rscParseRequest(RedisModuleString **argv, int argc) {
...
+    // Note: currently there is only one single case. For extending those cases we should use a trie here.
+    if(strcasestr(req->queryString, "KNN")) {
+      prepareOptionalTopKCase(req, argv, argc, status);
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply coord/src/module.c:626]: fail with a syntax error, same as shard or 2.2.9+ versions

File: src/query_parser/v1/parser.y [hunk 7]
```diff
@@ -0,0 +1,578 @@
+%left LOWEST.
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/query_parser/v1/parser.y:1]: yes, this is 2.2.9 parser, modified only where it is needed to use current functions

**Comment by MeirShpilraien** [comment]
@DvirDukhan general comment, can you add an explanation of the tests changes to the top comment. I believe all the tests should run with both dialect. If there are tests that should only run with one of the dialect we should have some `skip_on_dialect(0)` function that will skip the test if needed or will do different checks.

File: tests/pytests/test.py [hunk 8]
```diff
@@ -554,22 +554,22 @@ def testExplain(env):
     env.assertEqual(expected, res)
 
     q = '(hello world) "what what" hello|world @bar:[10 100]|@bar:[200 300]'
-    res = r.execute_command('ft.explain', 'idx', q)
+    res = r.execute_command('ft.explain', 'idx', q, 'DIALECT', 2)
```
Comment (MeirShpilraien) [inline tests/pytests/test.py:557]: Look at the comment I added to the PR, in my opinion all tests should run with both dialects.

File: coord/src/module.c (hunk 2 above)
Comment (MeirShpilraien) [reply coord/src/module.c]: @DvirDukhan , I see more such comment...

File: tests/pytests/test_dialect.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    env.expect("FT.CONFIG SET DEFAULT_DIALECT -1").error()
+    env.expect("FT.CONFIG SET DEFAULT_DIALECT 3").error()
+
+def test_dialect_query_errors(env):
+    conn = getConnectionByEnv(env)
+    env.expect("FT.CREATE idx SCHEMA t TEXT").ok()
+    conn.execute_command("HSET", "h", "t", "hello")
+    env.expect("FT.SEARCH idx 'hello' DIALECT").error().contains("Need argument for DIALECT")
+    env.expect("FT.SEARCH idx 'hello' DIALECT 0").error().contains("DIALECT requires a non negative integer >=1 and <= 2")
+    env.expect("FT.SEARCH idx 'hello' DIALECT 3").error().contains("DIALECT requires a non negative integer >=1 and <= 2")
```
[the remark this answers is not in the document]
Comment (oshadmi) [reply tests/pytests/test_dialect.py:32]: Using `FT.SEARCH` Vector query and `DIALECT` 1 will issue a plain syntax error with no additional explanation that it requires dialect 2, right? OK.
So thisscneario with `FT.SEARCH` is in cpp test.
I see `FT.EXPLAIN` is in py test with such syntax error failure.
Seems better to put them together (all in cpp, or all in py), no?
What about FT.AGGREGATE? Is it tested for this plain syntax error?

**Review by MeirShpilraien — COMMENTED** [review]
Main concerns:
1. Tests should all run with both dialects
2. Coverage report shows uncovered code

**Review by oshadmi — COMMENTED** [review]
One comment

**Review by oshadmi — APPROVED** [review]
Looks great 👍🏼

**Review by MeirShpilraien — APPROVED** [review]

**Pull request merged**